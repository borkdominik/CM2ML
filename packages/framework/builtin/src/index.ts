import { ArchimateParser } from '@cm2ml/archimate'
import { EcoreParser } from '@cm2ml/ecore'
import { GraphEncoder } from '@cm2ml/graph-encoder'
import type { GraphModel } from '@cm2ml/ir'
import { type Plugin, batchedCompose, compose } from '@cm2ml/plugin'
import { OneHotEncoder } from '@cm2ml/one-hot-encoder'
import { type Plugin, compose } from '@cm2ml/plugin'
import { TreeEncoder } from '@cm2ml/tree-encoder'
import { UmlParser } from '@cm2ml/uml'

export * from '@cm2ml/archimate'
export * from '@cm2ml/ecore'
export * from '@cm2ml/graph-encoder'
export * from '@cm2ml/one-hot-encoder'
export * from '@cm2ml/tree-encoder'
export * from '@cm2ml/uml'
export * from '@cm2ml/xmi-parser'

export type Parser = Plugin<string, GraphModel, any>

export const parsers: Parser[] = [ArchimateParser, EcoreParser, UmlParser]

export const parserMap = {
  [ArchimateParser.name]: ArchimateParser,
  [EcoreParser.name]: EcoreParser,
  [UmlParser.name]: UmlParser,
}

export type EncoderMetadata = any

export type Encoder = Plugin<GraphModel, unknown, any, unknown, EncoderMetadata>

export const encoders: Encoder[] = [GraphEncoder, TreeEncoder, OneHotEncoder]

export const encoderMap = {
  [GraphEncoder.name]: GraphEncoder,
  [TreeEncoder.name]: TreeEncoder,
  [OneHotEncoder.name]: OneHotEncoder,
}

export const plugins = parsers.flatMap((parser) =>
  encoders.map((encoder) => compose(parser, encoder)),
)

export const batchedPlugins = parsers.flatMap((parser) =>
  encoders.map((encoder) => batchedCompose(parser, encoder)),
)
