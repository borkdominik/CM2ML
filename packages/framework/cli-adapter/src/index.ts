import { existsSync } from 'node:fs'
import process from 'node:process'

import type { ExecutionError, METADATA_KEY, Parameter, ParameterMetadata, Plugin } from '@cm2ml/plugin'
import { getTypeConstructor } from '@cm2ml/plugin'
import type { PluginAdapterConfiguration } from '@cm2ml/plugin-adapter'
import { PluginAdapter } from '@cm2ml/plugin-adapter'
import { getMessage } from '@cm2ml/utils'
import { Stream } from '@yeger/streams'
import type { Command } from 'cac'
import { cac } from 'cac'

import { batchedPluginActionHandler } from './batched-plugin-action-handler'
import { getFeatureMetadataFromFile } from './feature-metadata-extractor'
import { pluginActionHandler } from './plugin-action-handler'

class CLI extends PluginAdapter<string, PluginAdapterConfiguration> {
  private readonly cli = cac('cm2ml')
    .option(
      '--out <file/directory>',
      'Path to output file (or directory for non-merged batch mode)',
      { type: [undefinedAwareConstructor(String)] },
    )
    .option('--pretty', 'Pretty print JSON output', { default: false, type: [Boolean] })

  protected onApply<Out, Parameters extends ParameterMetadata>(
    plugin: Plugin<string, Out, Parameters>,
  ) {
    const command = this.cli.command(`${plugin.name} <inputFile>`)
    registerCommandOptions(command, plugin.parameters)
    command.action((inputFile: string, options: Record<string, unknown>) =>
      pluginActionHandler(plugin, inputFile, options),
    )
  }

  protected onApplyBatched<Out, Parameters extends ParameterMetadata>(plugin: Plugin<string[], { data: (Out | ExecutionError)[], [METADATA_KEY]: unknown }, Parameters>) {
    const command = this.cli.command(`${plugin.name} <inputDir>`)
    registerCommandOptions(command, plugin.parameters)
    command.option('--merge', 'Merge all results into a single output', { default: false, type: [Boolean] })
    command.option('--start <start>', 'Index of the first model to encode', { default: undefined, type: [undefinedAwareConstructor(Number)] })
    command.option('--limit <limit>', 'Maximum number of models to encode', { default: undefined, type: [undefinedAwareConstructor(Number)] })
    command.action((inputDir: string, options: Record<string, unknown>) =>
      batchedPluginActionHandler(plugin, inputDir, options),
    )
  }

  protected onStart() {
    try {
      this.cli.help()
      this.cli.command('').action(() => this.cli.outputHelp())
      this.cli.parse()
    } catch (error) {
      console.error(getMessage(error))
      process.exit(1)
    }
  }
}

function registerCommandOptions<Parameters extends ParameterMetadata>(
  command: Command,
  parameters: Parameters,
) {
  Stream.fromObject(parameters).forEach(([name, parameter]) => {
    if (parameter.type === 'array<string>') {
      command.option(
          `--${createOptionName(name)} <${name}>`,
          createOptionDescription(parameter),
          {
            // type: [getTypeConstructor('array<string>')],
          },
      )
      return
    }
    if (parameter.type === 'string' && ['nodeFeatures', 'edgeFeatures'].includes(name)) {
      // These parameters are often too large for shell arguments, so we have to use a file instead
      command.option(
          `--${createOptionName(name)} <${name}File>`,
          createOptionDescription(parameter),
          {
            default: parameter.defaultValue,
            type: [(input: unknown) => {
              if (!input) {
                return parameter.defaultValue
              }
              const inputFile = String(input)
              if (!inputFile) {
                return parameter.defaultValue
              }
              try {
                // check if the file exists
                if (!existsSync(inputFile)) {
                  return inputFile
                }
              } catch (error) {
                return inputFile
              }
              return getFeatureMetadataFromFile(inputFile, `__metadata__.${name}`)
            }],
          },
      )
      return
    }
    if (parameter.type !== 'boolean') {
      command.option(
          `--${createOptionName(name)} <${name}>`,
          createOptionDescription(parameter),
          {
            default: parameter.defaultValue,
            type: [getTypeConstructor(parameter.type)],
          },
      )
      return
    }
    if (parameter.defaultValue !== true) {
      command.option(
          `--${createOptionName(name)}`,
          createOptionDescription(parameter),
          {
            default: parameter.defaultValue,
            type: [getTypeConstructor(parameter.type)],
          },
      )
    }
    if (parameter.defaultValue !== false) {
      const alteredDescription =
          parameter.description.substring(0, 1).toLowerCase() +
          parameter.description.substring(1)
      command.option(
          `--no-${createOptionName(name)}`,
          `Do not ${alteredDescription}`,
          {
            default: parameter.defaultValue,
            type: [getTypeConstructor(parameter.type)],
          },
      )
    }
  })
}

function createOptionName(parameterName: string) {
  return parameterName
    .split(/\.?(?=[A-Z])/)
    .join('-')
    .toLowerCase()
}

function createOptionDescription(parameter: Parameter) {
  if (parameter.type !== 'string') {
    return parameter.description
  }
  const allowedValues = parameter.allowedValues
  if (!allowedValues || allowedValues.length === 0) {
    return parameter.description
  }
  return `${parameter.description} (options: ${allowedValues.join(', ')})`
}

function undefinedAwareConstructor(constructor: NumberConstructor | StringConstructor | ((input: unknown) => boolean)) {
  return (input: unknown) => input === undefined ? undefined : constructor(input)
}

export function createCLI() {
  return new CLI()
}
