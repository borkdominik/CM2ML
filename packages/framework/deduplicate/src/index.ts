import type { StructuredOutput } from '@cm2ml/plugin'
import { ExecutionError, definePlugin } from '@cm2ml/plugin'
import deepEqual from 'fast-deep-equal'

export const DUPLICATE_SYMBOL = Symbol('DUPLICATE')

export type DuplicateSymbol = typeof DUPLICATE_SYMBOL

const name = 'deduplicate'

export function deduplicate<Data, Metadata>() {
  return definePlugin({
    name,
    parameters: {
      deduplicate: {
        type: 'boolean',
        defaultValue: false,
        description: 'Remove duplicate results',
      },
      deduplicationData: {
        type: 'list<string>',
        unique: false,
        ordered: false,
        defaultValue: [],
        description: 'Additional data to check for duplicates. Must be a serialization of deduplicated batch output.',
        processFile: (content: string) => content,
      },
    },
    invoke(batch: StructuredOutput<(Data | ExecutionError)[], Metadata>, { deduplicate, deduplicationData }) {
      if (!deduplicate) {
        return batch
      }
      return removeDuplicates(batch, deduplicationData.flatMap((data) => extractDeduplicationData(data)))
    },
  })
}

function extractDeduplicationData(additionalData: string): unknown[] {
  if (additionalData === '') {
    return []
  }
  const parsed: unknown = JSON.parse(additionalData)
  if (Array.isArray(parsed)) {
    return parsed
  }
  if (typeof parsed !== 'object' || parsed === null) {
    throw new ExecutionError('Invalid duplication data', name)
  }
  if (!('data' in parsed)) {
    throw new ExecutionError('Invalid duplication data', name)
  }
  const data: unknown = parsed.data
  if (Array.isArray(data)) {
    return data
  }
  if (typeof data !== 'object' || data === null) {
    throw new ExecutionError('Invalid duplication data', name)
  }
  return Object.values(data)
}

function removeDuplicates<Data, Metadata>(batch: StructuredOutput<(Data | ExecutionError)[], Metadata>, additionalData: unknown[]): StructuredOutput<(Data | ExecutionError | DuplicateSymbol)[], Metadata> {
  const duplicateIndices = new Set<number>()
  const output: (Data | ExecutionError | DuplicateSymbol)[] = []
  const data = batch.data
  additionalData.forEach((additionalEntry) => {
    data.forEach((entry, index) => {
      if (deepEqual(additionalEntry, entry)) {
        duplicateIndices.add(index)
      }
    })
  })
  data.forEach((entry, index) => {
    if (duplicateIndices.has(index)) {
      output.push(DUPLICATE_SYMBOL)
      return
    }
    output.push(entry)
    if (entry instanceof ExecutionError) {
      return
    }
    for (let otherIndex = index + 1; otherIndex < data.length; otherIndex++) {
      if (duplicateIndices.has(otherIndex)) {
        continue
      }
      const otherEntry = data[otherIndex]!
      if (otherEntry instanceof ExecutionError) {
        continue
      }
      if (deepEqual(entry, otherEntry)) {
        duplicateIndices.add(otherIndex)
      }
    }
  })
  return { data: output, metadata: batch.metadata }
}
