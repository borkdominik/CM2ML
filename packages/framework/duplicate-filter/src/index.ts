import type { StructuredOutput } from '@cm2ml/plugin'
import { ExecutionError, definePlugin } from '@cm2ml/plugin'
import deepEqual from 'fast-deep-equal'

export const DUPLICATE_SYMBOL = Symbol('DUPLICATE')

export type DuplicateSymbol = typeof DUPLICATE_SYMBOL

const name = 'duplicate-filter'

export function filterDuplicates<Data, Metadata>() {
  return definePlugin({
    name,
    parameters: {
      removeDuplicates: {
        type: 'boolean',
        defaultValue: false,
        description: 'Remove duplicate results',
      },
    },
    invoke(batch: StructuredOutput<(Data | ExecutionError)[], Metadata>, parameters) {
      if (!parameters.removeDuplicates) {
        return batch
      }
      return removeDuplicates(batch)
    },
  })
}

function removeDuplicates<Data, Metadata>(batch: StructuredOutput<(Data | ExecutionError)[], Metadata>): StructuredOutput<(Data | ExecutionError | DuplicateSymbol)[], Metadata> {
  const duplicateIndices = new Set<number>()
  const output: (Data | ExecutionError | DuplicateSymbol)[] = []
  const data = batch.data
  data.forEach((entry, index) => {
    if (duplicateIndices.has(index)) {
      output.push(DUPLICATE_SYMBOL)
      return
    }
    output.push(entry)
    if (entry instanceof ExecutionError) {
      return
    }
    for (let otherIndex = index + 1; otherIndex < data.length; otherIndex++) {
      if (duplicateIndices.has(otherIndex)) {
        continue
      }
      const otherEntry = data[otherIndex]!
      if (otherEntry instanceof ExecutionError) {
        continue
      }
      if (deepEqual(entry, otherEntry)) {
        duplicateIndices.add(otherIndex)
      }
    }
  })
  return { data: output, metadata: batch.metadata }
}
