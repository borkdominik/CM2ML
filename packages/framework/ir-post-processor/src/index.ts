import type { GraphModel, GraphNode } from '@cm2ml/ir'
import { definePlugin } from '@cm2ml/plugin'
import { Stream } from '@yeger/streams'

import { validateModel } from './validations'

export const IrPostProcessor = definePlugin({
  name: 'ir-post-processor',
  parameters: {
  },
  invoke: (model: GraphModel) => {
    if (!model.settings.strict) {
      removeInvalidNodesFromModel(model)
    }
    validateModel(model)
    return model
  },
})

function removeInvalidNodesFromModel(model: GraphModel) {
  Stream.from(model.nodes)
    .filter(isNodeInvalid)
    .forEach((node) => {
      model.removeNode(node)
    })
}

function isNodeInvalid(node: GraphNode) {
  function isNodeUnconnected() {
    return node.incomingEdges.size === 0 && node.outgoingEdges.size === 0
  }
  function isNodeUnidentifiable() {
    return node.id === undefined
  }
  if (node === node.model.root) {
    return false
  }
  return isNodeUnconnected() || isNodeUnidentifiable()
}
