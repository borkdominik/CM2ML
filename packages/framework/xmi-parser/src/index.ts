import type { Attribute, GraphNode, Settings, Value } from '@cm2ml/ir'
import { GraphModel } from '@cm2ml/ir'
import { definePlugin } from '@cm2ml/plugin'
import { parseNamespace } from '@cm2ml/utils'
import { Stream } from '@yeger/streams'
import { Element, Text } from 'domhandler'
import type { Document, Node } from 'domhandler'
import { parseDocument } from 'htmlparser2'

export type TextNodeHandler = (node: GraphNode, text: string) => void

export function createXmiParser(
  idAttribute: string,
  textNodeHandler: TextNodeHandler,
) {
  return definePlugin({
    name: 'xmi',
    parameters: {
      debug: {
        type: 'boolean',
        defaultValue: false,
        description: 'Log debug information and validate results if strict is also enabled.',
      },
      strict: {
        type: 'boolean',
        defaultValue: false,
        description: 'Fail when encountering unknown or invalid input.',
      },
    },
    invoke: (input: string, settings) =>
      parse(input, { ...settings, idAttribute }, textNodeHandler),
  })
}

function parse(
  xmi: string,
  settings: Settings,
  textNodeHandler: TextNodeHandler,
): GraphModel {
  const document = parseDocument(xmi, {
    xmlMode: true,
  })
  return mapDocument(document, settings, textNodeHandler)
}

function mapDocument(
  document: Document,
  settings: Settings,
  textNodeHandler: TextNodeHandler,
) {
  const elementChildren = Stream.from(document.childNodes)
    .map((node) => (isElement(node) ? node : null))
    .filterNonNull()
    .toArray()
  const root = elementChildren[0]
  if (elementChildren.length !== 1 || !root) {
    throw new Error('Expected exactly one root element')
  }
  const model = new GraphModel(settings, root.tagName)
  initNodeFromElement(model.root, root, textNodeHandler)
  return model
}

function createNodeFromElement(
  model: GraphModel,
  element: Element,
  textNodeHandler: TextNodeHandler,
): GraphNode {
  const node = model.addNode(element.tagName)
  initNodeFromElement(node, element, textNodeHandler)
  return node
}

function initNodeFromElement(
  node: GraphNode,
  element: Element,
  textNodeHandler: TextNodeHandler,
) {
  Stream.fromObject(element.attribs)
    .map(mapAttribute)
    .forEach((attribute) => node.addAttribute(attribute, true))
  Stream.from(element.childNodes).forEach((child) => {
    if (isElement(child)) {
      const childNode = createNodeFromElement(
        node.model,
        child,
        textNodeHandler,
      )
      node.addChild(childNode)
      return
    }
    if (isText(child)) {
      const text = child.data.trim()
      if (text === '') {
        return
      }
      textNodeHandler(node, text)
    }
  })
}

function mapAttribute([name, value]: [string, string]): Attribute {
  const xmiValue = mapValue(value)
  const parsedName = parseNamespace(name)
  if (typeof parsedName === 'object') {
    return {
      name,
      simpleName: parsedName.name,
      namespace: parsedName.namespace,
      value: xmiValue,
    }
  }
  return { name, value: xmiValue }
}

function mapValue(value: string): Value {
  if (!value.includes(':') || value.startsWith('http')) {
    return { literal: value }
  }
  const [namespace, ...rest] = value.split(':')
  return { literal: rest.join(':'), namespace }
}

function isElement(node: Node): node is Element {
  return node.type === 'tag' || node instanceof Element
}

function isText(node: Node): node is Text {
  return node.type === 'text' || node instanceof Text
}
