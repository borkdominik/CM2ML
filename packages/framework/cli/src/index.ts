import { batchedPlugins, plugins } from '@cm2ml/builtin'
import { createCLI } from '@cm2ml/cli-adapter'

export function startPreConfiguredCli() {
  createCLI()
    .applyAll(batchedPlugins, { batched: true })
    .applyAll(plugins, { batched: false })
    .start()
}
