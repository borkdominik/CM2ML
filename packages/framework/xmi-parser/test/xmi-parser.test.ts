import fs from 'node:fs'

import { describe, expect, it } from 'vitest'

import { createXmiParser } from '../src/index'

describe('xmi-parser', () => {
  it('should parse', () => {
    const input = fs.readFileSync('../../../models/xmi-model.xmi', 'utf-8')
    const xmiParser = createXmiParser('id', () => {})
    const output = xmiParser.invoke(input, {
      debug: false,
      strict: true,
    })
    expect(output.show()).toMatchSnapshot()
  })
})
