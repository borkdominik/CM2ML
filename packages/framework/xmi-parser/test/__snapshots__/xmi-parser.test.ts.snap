// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`xmi-parser > should parse 1`] = `
"<xmi:XMI xmlns:mofext="http://www.omg.org/spec/MOF/20131001" xmlns:uml="http://www.omg.org/spec/UML/20131001" xmlns:xmi="http://www.omg.org/spec/XMI/20131001">
  <uml:Package URI="http://www.omg.org/spec/XMI/20131001" name="XMI" xmi:id="_XMI" xmi:type="Package">
    <packagedElement name="Delete" xmi:id="_XMI-Delete" xmi:type="Class">
      <generalization xmi:id="_XMI-Delete-Generalization.0" xmi:type="Generalization">
        <general xmi:idref="_XMI-Difference" />
      </generalization>
    </packagedElement>
    <packagedElement name="A_addition_add" xmi:id="_XMI-A_addition_add" xmi:type="Association">
      <memberEnd xmi:idref="_XMI-A_addition_add-add" />
      <memberEnd xmi:idref="_XMI-Add-addition" />
      <ownedEnd name="add" visibility="public" xmi:id="_XMI-A_addition_add-add" xmi:type="Property">
        <type xmi:idref="_XMI-Add" />
        <association xmi:idref="_XMI-A_addition_add" />
        <upperValue value="*" xmi:id="_XMI-A_addition_add-add-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-A_addition_add-add-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedEnd>
    </packagedElement>
    <packagedElement name="Documentation" xmi:id="_XMI-Documentation" xmi:type="Class">
      <ownedAttribute name="contact" visibility="public" xmi:id="_XMI-Documentation-contact" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <lowerValue xmi:id="_XMI-Documentation-contact-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="exporter" visibility="public" xmi:id="_XMI-Documentation-exporter" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <lowerValue xmi:id="_XMI-Documentation-exporter-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="exporterVersion" visibility="public" xmi:id="_XMI-Documentation-exporterVersion" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <lowerValue xmi:id="_XMI-Documentation-exporterVersion-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="exporterID" visibility="public" xmi:id="_XMI-Documentation-exporterID" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <lowerValue xmi:id="_XMI-Documentation-exporterID-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="longDescription" visibility="public" xmi:id="_XMI-Documentation-longDescription" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <upperValue value="*" xmi:id="_XMI-Documentation-longDescription-LiteralUnlimitedNatural.1" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Documentation-longDescription-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="shortDescription" visibility="public" xmi:id="_XMI-Documentation-shortDescription" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <upperValue value="*" xmi:id="_XMI-Documentation-shortDescription-LiteralUnlimitedNatural.1" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Documentation-shortDescription-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="notice" visibility="public" xmi:id="_XMI-Documentation-notice" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <upperValue value="*" xmi:id="_XMI-Documentation-notice-LiteralUnlimitedNatural.1" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Documentation-notice-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="owner" visibility="public" xmi:id="_XMI-Documentation-owner" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <upperValue value="*" xmi:id="_XMI-Documentation-owner-LiteralUnlimitedNatural.1" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Documentation-owner-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="timestamp" visibility="public" xmi:id="_XMI-Documentation-timestamp" xmi:type="Property">
        <type xmi:idref="_XMI-DateTime" />
        <lowerValue xmi:id="_XMI-Documentation-timestamp-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
    </packagedElement>
    <packagedElement name="A_container_difference" xmi:id="_XMI-A_container_difference" xmi:type="Association">
      <memberEnd xmi:idref="_XMI-Difference-container" />
      <memberEnd xmi:idref="_XMI-Difference-difference" />
    </packagedElement>
    <packagedElement name="A_target_difference" xmi:id="_XMI-A_target_difference" xmi:type="Association">
      <memberEnd xmi:idref="_XMI-A_target_difference-difference" />
      <memberEnd xmi:idref="_XMI-Difference-target" />
      <ownedEnd name="difference" visibility="public" xmi:id="_XMI-A_target_difference-difference" xmi:type="Property">
        <type xmi:idref="_XMI-Difference" />
        <association xmi:idref="_XMI-A_target_difference" />
        <upperValue value="*" xmi:id="_XMI-A_target_difference-difference-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-A_target_difference-difference-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedEnd>
    </packagedElement>
    <packagedElement isAbstract="true" name="Difference" xmi:id="_XMI-Difference" xmi:type="Class">
      <ownedAttribute name="container" visibility="public" xmi:id="_XMI-Difference-container" xmi:type="Property">
        <type xmi:idref="_XMI-Difference" />
        <association xmi:idref="_XMI-A_container_difference" />
        <lowerValue xmi:id="_XMI-Difference-container-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute aggregation="composite" name="difference" visibility="public" xmi:id="_XMI-Difference-difference" xmi:type="Property">
        <type xmi:idref="_XMI-Difference" />
        <association xmi:idref="_XMI-A_container_difference" />
        <upperValue value="*" xmi:id="_XMI-Difference-difference-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Difference-difference-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="target" visibility="public" xmi:id="_XMI-Difference-target" xmi:type="Property">
        <association xmi:idref="_XMI-A_target_difference" />
        <type href="http://www.omg.org/spec/UML/20131001/UML.xmi#Element" />
        <upperValue value="*" xmi:id="_XMI-Difference-target-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Difference-target-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedAttribute>
    </packagedElement>
    <packagedElement name="Extension" xmi:id="_XMI-Extension" xmi:type="Class">
      <ownedAttribute name="extender" visibility="public" xmi:id="_XMI-Extension-extender" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
      </ownedAttribute>
      <ownedAttribute name="extenderID" visibility="public" xmi:id="_XMI-Extension-extenderID" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#String" />
        <lowerValue xmi:id="_XMI-Extension-extenderID-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
    </packagedElement>
    <packagedElement name="XMI" xmi:id="_XMI-XMI" xmi:type="Class">
      <ownedAttribute name="documentation" visibility="public" xmi:id="_XMI-XMI-documentation" xmi:type="Property">
        <type xmi:idref="_XMI-Documentation" />
        <lowerValue xmi:id="_XMI-XMI-documentation-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="difference" visibility="public" xmi:id="_XMI-XMI-difference" xmi:type="Property">
        <type xmi:idref="_XMI-Difference" />
        <upperValue value="*" xmi:id="_XMI-XMI-difference-LiteralUnlimitedNatural.1" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-XMI-difference-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="extension" visibility="public" xmi:id="_XMI-XMI-extension" xmi:type="Property">
        <type xmi:idref="_XMI-Extension" />
        <upperValue value="*" xmi:id="_XMI-XMI-extension-LiteralUnlimitedNatural.1" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-XMI-extension-LiteralInteger.2" xmi:type="LiteralInteger" />
      </ownedAttribute>
    </packagedElement>
    <packagedElement name="Replace" xmi:id="_XMI-Replace" xmi:type="Class">
      <generalization xmi:id="_XMI-Replace-Generalization.0" xmi:type="Generalization">
        <general xmi:idref="_XMI-Difference" />
      </generalization>
      <ownedAttribute name="position" visibility="public" xmi:id="_XMI-Replace-position" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#Integer" />
        <lowerValue xmi:id="_XMI-Replace-position-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="replacement" visibility="public" xmi:id="_XMI-Replace-replacement" xmi:type="Property">
        <association xmi:idref="_XMI-A_replacement_replace" />
        <type href="http://www.omg.org/spec/UML/20131001/UML.xmi#Element" />
        <upperValue value="*" xmi:id="_XMI-Replace-replacement-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Replace-replacement-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedAttribute>
    </packagedElement>
    <packagedElement name="DateTime" xmi:id="_XMI-DateTime" xmi:type="PrimitiveType" />
    <packagedElement name="A_replacement_replace" xmi:id="_XMI-A_replacement_replace" xmi:type="Association">
      <memberEnd xmi:idref="_XMI-A_replacement_replace-replace" />
      <memberEnd xmi:idref="_XMI-Replace-replacement" />
      <ownedEnd name="replace" visibility="public" xmi:id="_XMI-A_replacement_replace-replace" xmi:type="Property">
        <type xmi:idref="_XMI-Replace" />
        <association xmi:idref="_XMI-A_replacement_replace" />
        <upperValue value="*" xmi:id="_XMI-A_replacement_replace-replace-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-A_replacement_replace-replace-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedEnd>
    </packagedElement>
    <packagedElement name="Add" xmi:id="_XMI-Add" xmi:type="Class">
      <generalization xmi:id="_XMI-Add-Generalization.0" xmi:type="Generalization">
        <general xmi:idref="_XMI-Difference" />
      </generalization>
      <ownedAttribute name="position" visibility="public" xmi:id="_XMI-Add-position" xmi:type="Property">
        <type href="http://www.omg.org/spec/UML/20131001/PrimitiveTypes.xmi#Integer" />
        <lowerValue xmi:id="_XMI-Add-position-LiteralInteger.1" xmi:type="LiteralInteger" />
      </ownedAttribute>
      <ownedAttribute name="addition" visibility="public" xmi:id="_XMI-Add-addition" xmi:type="Property">
        <association xmi:idref="_XMI-A_addition_add" />
        <type href="http://www.omg.org/spec/UML/20131001/UML.xmi#Element" />
        <upperValue value="*" xmi:id="_XMI-Add-addition-LiteralUnlimitedNatural.2" xmi:type="LiteralUnlimitedNatural" />
        <lowerValue xmi:id="_XMI-Add-addition-LiteralInteger.3" xmi:type="LiteralInteger" />
      </ownedAttribute>
    </packagedElement>
  </uml:Package>
  <mofext:Tag name="org.omg.xmi.nsPrefix" value="xmi" xmi:id="_org.omg.xmi.nsPrefix" xmi:type="Tag">
    <element xmi:idref="_XMI" />
  </mofext:Tag>
  <mofext:Tag name="org.omg.xmi.superClassFirst" value="true" xmi:id="_org.omg.xmi.superClassFirst" xmi:type="Tag">
    <element xmi:idref="_XMI" />
  </mofext:Tag>
  <mofext:Tag name="org.omg.xmi.useSchemaExtension" value="true" xmi:id="_org.omg.xmi.useSchemaExtension" xmi:type="Tag">
    <element xmi:idref="_XMI" />
  </mofext:Tag>
  <mofext:Tag name="org.omg.xmi.element" value="true" xmi:id="_org.omg.xmi.element" xmi:type="Tag">
    <element xmi:idref="_XMI" />
  </mofext:Tag>
  <mofext:Tag name="org.omg.xmi.attribute" value="false" xmi:id="_org.omg.xmi.attribute" xmi:type="Tag">
    <element xmi:idref="_XMI" />
  </mofext:Tag>
  <mofext:Tag name="org.omg.xmi.schemaType" value="http://www.w3.org/2001/XMLSchema#dateTime" xmi:id="_org.omg.xmi.schemaType" xmi:type="Tag">
    <element xmi:idref="_XMI-DateTime" />
  </mofext:Tag>
</xmi:XMI>"
`;
